### ISSUE TITLE
TODO

### ISSUE THREAD

**Issue by pdasigi** [pdasigi/eqqa#1]
High priority:
- [ ] Look into MOCHA and finalize how much we can use it for evaluation.
- [ ] If we decide we need additional human annotations like MOCHA, finalize the annotation scheme.
- [ ] Finalize OOD evaluation sets for each of the datasets we are training quality estimators for (MS MARCO, NarrativeQA, Qasper)

Medium priority:
- [ ] Augment training data for the (question, context, prediction) -> F1 score model and see if we can do better on Qasper
- [ ] LSTM with attention on final layer features -> F1 score for Qasper 
- [ ] Train QCP -> F1 model on MS MARCO and NarrativeQA

Low priority:
- [ ] Include LM score as an additional feature in the MLP quality estimator

After we have an evaluation setup: 
- [ ] Compare different target metrics for quality estimation (e.g.: Is regressing against ROUGE better than doing it against F1?): Evaluation metric will be rank correlation (possibly binned) against human annotations.
- [ ] Compare multi-tasking on regressing against metrics with calibrating on individual metrics. Evaluated based on rank correlation against human annotations.

**Comment by PastelBelem8** [comment]
As agreed during our discussion, I will focus on the implementation of Tree Regression on Qasper MLP quality estimator.


In that regards, I found a few papers that describe a differentiable and softer/smoother version of decision trees. The first paper to research differentiable decision trees seems to date back to [Suarez and Lusko (1999)](https://ieeexplore.ieee.org/stamp/stamp.jsp?tp=&arnumber=817409). 

Numerous papers have appeared after that one, for example, the paper [Soft Decision Trees](https://www.cs.cornell.edu/~oirsoy/files/icpr21.pdf) (Irsoy et al. 2012). While I don't know yet the exact differences to Suarez and Lusko (1999), Irsoy et al. 2012 is a 4-page paper summarizing the main changes from an implementation point of view and contains the pseudocode algorithm pseudo-code. The gist seems to be using gradient descent and batches to build the trees incrementally - one step tree level at a time. Each example can follow multiple paths rather than a single one (hence, the name).


This smoother version can be incorporated into the MLP model in a bi-level optimization fashion as follows: 
```python
while not_converged:
   regression_input = build_regression_input(dev_data)  // create the regression inputs
   soft_tree = build_soft_tree(regression_input, target_metric) // build soft decision tree
   

   
```
where, at each iteration, we first fit a Soft-Decision Tree based on the embeddings of the LM at that iteration, then in the model part, we use this Soft Decision Tree to provide the calibrated scores and thus guide the updates of the LM based on these values.)
The idea of Soft Decision Trees has been recently used to propose differentiable “layers” or “components” in recent works:
[Adaptive Neural Trees](https://arxiv.org/pdf/1807.06699.pdf) (Tanno 2019) ([code in Python 2.7](https://github.com/rtanno21609/AdaptiveNeuralTrees))
[The Tree Ensemble Layer](https://arxiv.org/pdf/2002.07772.pdf) (Hazimeh et al. 2020 ICML) ([code for TensorFlow](https://github.com/google-research/google-research/tree/master/tf_trees)) from GoogleResearch proposes a tree-based ensemble layer and has some TF code available.


I found several PyTorch implementations ([example1](https://github.com/xuyxu/Soft-Decision-Tree), [example2](https://github.com/kimhc6028/soft-decision-tree)) of soft decision trees (also called differentiable decision trees). These implementations seem to be highly influenced by Google’s paper: [Distilling a Neural Network Into a Soft Decision Tree](https://arxiv.org/pdf/1711.09784.pdf) (2017).
There’s also this blogpost on [how to build a Soft Decision Tree in TF](https://towardsdatascience.com/building-a-decision-tree-in-tensorflow-742438cb483e).

**Comment by PastelBelem8** [comment]
For the evaluation setup, the following papers can be insightful for the creation of an Answer Equivalence task: 
- [Tomayto, Tomahto. Beyond Token-level Answer Equivalence for Question Answering Evaluation (Bulian et al 2022)](https://arxiv.org/abs/2202.07654): propose an Answer Equivalence (AE) task and corresponding SQUAD-based human annotated dataset. They advocate for assymetric metrics, which reflect partially correct/incorrect, missing information, extra information categories in the annotation procedure. 
- [MOCHA: A Dataset for Training and Evaluating Generative Reading Comprehension Metrics (Chen et al. 2020)](https://arxiv.org/abs/2010.03636): repurpose different MC and extractive settings to be generative and collect human annotations for equivalent answers in a 5-point likert scale.

**Comment by PastelBelem8** [comment]
Another idea for QE could be to calibrate the final model score by fitting a model that uses the values of the different metrics to compute the model confidence... Since each metric can have different strengths or weaknesses, combining the different models (e.g., via linear combination)
